In [1]:
import mysql.connector
import config

connection = mysql.connector.connect(**config.connection())

In [2]:
def get_data(query):
    '''This function returns the data from the database'''
    cursor = connection.cursor()
    cursor.execute(query)
    data = cursor.fetchall()
    return data

In [3]:
print(get_data(
    '''SHOW DATABASES;'''
))

[('employees',), ('information_schema',), ('mysql',), ('performance_schema',), ('quizdata',), ('sys',)]


In [33]:
get_data('''SELECT * FROM employees LIMIT 10;''')

[(100,
  datetime.date(1977, 1, 1),
  'John',
  'Doe',
  'M',
  datetime.date(2000, 1, 1)),
 (10001,
  datetime.date(1953, 9, 2),
  'Georgi',
  'Facello',
  'M',
  datetime.date(1986, 6, 26)),
 (10002,
  datetime.date(1964, 6, 2),
  'Bezalel',
  'Simmel',
  'F',
  datetime.date(1985, 11, 21)),
 (10003,
  datetime.date(1959, 12, 3),
  'Parto',
  'Bamford',
  'M',
  datetime.date(1986, 8, 28)),
 (10004,
  datetime.date(1954, 5, 1),
  'Chirstian',
  'Koblick',
  'M',
  datetime.date(1986, 12, 1)),
 (10005,
  datetime.date(1955, 1, 21),
  'Kyoichi',
  'Maliniak',
  'M',
  datetime.date(1989, 9, 12)),
 (10006,
  datetime.date(1953, 4, 20),
  'Anneke',
  'Preusig',
  'F',
  datetime.date(1989, 6, 2)),
 (10007,
  datetime.date(1957, 5, 23),
  'Tzvetan',
  'Zielinski',
  'F',
  datetime.date(1989, 2, 10)),
 (10008,
  datetime.date(1958, 2, 19),
  'Saniya',
  'Kalloufi',
  'M',
  datetime.date(1994, 9, 15)),
 (10009,
  datetime.date(1952, 4, 19),
  'Sumant',
  'Peac',
  'F',
  datetime.date(198

ASSIGN EMPLOYEE NUMBER 110022 AS A MANAGER TO ALL EMPLOYEES FROM 10001 TO 10020 AND EMPLOYEE 110039 AS A 
MANAGER TO ALL EMPLOYEES FROM 10021 TO 10040

In [34]:
get_data(
    '''
    SELECT 
        A.*
    FROM
        (SELECT 
            e.emp_no AS employee_ID,
                MIN(de.dept_no) AS department_code,
                (SELECT
                    emp_no
                FROM
                    dept_manager
                WHERE
                    emp_no = 110022) AS manager_ID
        FROM
            employees e
        JOIN dept_emp de ON e.emp_no = de.emp_no
        WHERE
            e.emp_no <= 10020
        GROUP BY e.emp_no
        ORDER BY e.emp_no) AS A
    UNION  
    SELECT 
        B.*
    FROM
        (SELECT 
            e.emp_no AS employee_ID,
                MIN(de.dept_no) AS department_code,
                (SELECT
                    emp_no
                FROM
                    dept_manager
                WHERE
                    emp_no = 110039) AS manager_ID
    FROM
        employees e
    JOIN dept_emp de ON e.emp_no = de.emp_no
    WHERE
        e.emp_no > 10020
    GROUP BY e.emp_no
    ORDER BY e.emp_no
    LIMIT 20) AS B;
    '''
)

[(10001, 'd005', 110022),
 (10002, 'd007', 110022),
 (10003, 'd004', 110022),
 (10004, 'd004', 110022),
 (10005, 'd003', 110022),
 (10006, 'd005', 110022),
 (10007, 'd008', 110022),
 (10008, 'd005', 110022),
 (10009, 'd006', 110022),
 (10010, 'd004', 110022),
 (10011, 'd009', 110022),
 (10012, 'd005', 110022),
 (10013, 'd003', 110022),
 (10014, 'd005', 110022),
 (10015, 'd008', 110022),
 (10016, 'd007', 110022),
 (10017, 'd001', 110022),
 (10018, 'd004', 110022),
 (10019, 'd008', 110022),
 (10020, 'd004', 110022),
 (10021, 'd005', 110039),
 (10022, 'd005', 110039),
 (10023, 'd005', 110039),
 (10024, 'd004', 110039),
 (10025, 'd005', 110039),
 (10026, 'd004', 110039),
 (10027, 'd005', 110039),
 (10028, 'd005', 110039),
 (10029, 'd004', 110039),
 (10030, 'd004', 110039),
 (10031, 'd005', 110039),
 (10032, 'd004', 110039),
 (10033, 'd006', 110039),
 (10034, 'd007', 110039),
 (10035, 'd004', 110039),
 (10036, 'd003', 110039),
 (10037, 'd005', 110039),
 (10038, 'd009', 110039),
 (10039, 'd0

CREATE A TABLE CALLED EMP_MANAGER

In [35]:
get_data(
    '''
    DROP TABLE IF EXISTS emp_manager;
    '''
)

[]

In [36]:
get_data(
    ''' CREATE TABLE IF NOT EXISTS emp_manager (
        emp_no INT NOT NULL,
        dept_no CHAR(4) NULL,
        manager_no INT NOT NULL
    );'''
)

[]

In [37]:
get_data(
    '''
INSERT INTO emp_manager
SELECT 
    u.*
FROM
    (SELECT 
        a.*
    FROM
        (SELECT 
        e.emp_no AS employee_ID,
            MIN(de.dept_no) AS department_code,
            (SELECT 
                    emp_no
                FROM
                    dept_manager
                WHERE
                    emp_no = 110022) AS manager_ID
    FROM
        employees e
    JOIN dept_emp de ON e.emp_no = de.emp_no
    WHERE
        e.emp_no <= 10020
    GROUP BY e.emp_no
    ORDER BY e.emp_no) AS a UNION SELECT 
        b.*
    FROM
        (SELECT 
        e.emp_no AS employee_ID,
            MIN(de.dept_no) AS department_code,
            (SELECT 
                    emp_no
                FROM
                    dept_manager
                WHERE
                    emp_no = 110039) AS manager_ID
    FROM
        employees e
    JOIN dept_emp de ON e.emp_no = de.emp_no
    WHERE
        e.emp_no > 10020
    GROUP BY e.emp_no
    ORDER BY e.emp_no
    LIMIT 20) AS b UNION SELECT 
        c.*
    FROM
        (SELECT 
        e.emp_no AS employee_ID,
            MIN(de.dept_no) AS department_code,
            (SELECT 
                    emp_no
                FROM
                    dept_manager
                WHERE
                    emp_no = 110039) AS manager_ID
    FROM
        employees e
    JOIN dept_emp de ON e.emp_no = de.emp_no
    WHERE
        e.emp_no = 110022
    GROUP BY e.emp_no) AS c UNION SELECT 
        d.*
    FROM
        (SELECT 
        e.emp_no AS employee_ID,
            MIN(de.dept_no) AS department_code,
            (SELECT 
                    emp_no
                FROM
                    dept_manager
                WHERE
                    emp_no = 110022) AS manager_ID
    FROM
        employees e
    JOIN dept_emp de ON e.emp_no = de.emp_no
    WHERE
        e.emp_no = 110039
    GROUP BY e.emp_no) AS d) as u;
    '''
)

[]

FROM THE EMP_MANAGER TABLE, EXTRACT THE RECORD DATA ONLY OF THOSE EMPLOYEES WHO ARE MANAGERS AS WELL.

In [38]:
man_employees = get_data(
    '''
    SELECT e1.*
    FROM
        emp_manager e1
    JOIN
        emp_manager e2 ON e1.emp_no = e2.manager_no
    WHERE
        e2.emp_no IN (
            SELECT manager_no FROM emp_manager
        );
    '''
)

In [39]:
print(man_employees)

[(110039, 'd001', 110022), (110022, 'd001', 110039)]


In [40]:
len(man_employees)

2

"VIEWS" ARE USED TO CREATE A TABLE FROM ANOTHER TABLE.

VIEWS IN SQL: 
[]: # Language: sql
[]: # Path: notebooks/queries.ipynb

VISUALIZE ONLY THE PERIOD ENCOMPASSING THE LAST CONTRACT OF EACH EMPLOYEE

In [41]:
last_contract = get_data(
    '''
    SELECT 
        emp_no, from_date, to_date,
        COUNT(emp_no) AS Num
    FROM
        dept_emp
    GROUP BY 
        emp_no
    HAVING Num > 1 LIMIT 20;
    '''
)

In [42]:
print(last_contract)

[(10010, datetime.date(1996, 11, 24), datetime.date(2000, 6, 26), 2), (10018, datetime.date(1992, 7, 29), datetime.date(9999, 1, 1), 2), (10029, datetime.date(1991, 9, 18), datetime.date(1999, 7, 8), 2), (10040, datetime.date(1993, 2, 14), datetime.date(2002, 1, 22), 2), (10050, datetime.date(1990, 12, 25), datetime.date(1992, 11, 5), 2), (10060, datetime.date(1989, 5, 28), datetime.date(1992, 11, 11), 2), (10070, datetime.date(1985, 10, 14), datetime.date(1995, 10, 18), 2), (10080, datetime.date(1994, 9, 28), datetime.date(1997, 7, 9), 2), (10088, datetime.date(1988, 9, 2), datetime.date(1992, 3, 21), 2), (10098, datetime.date(1985, 5, 13), datetime.date(1989, 6, 29), 2), (10108, datetime.date(1999, 12, 6), datetime.date(2001, 10, 20), 2), (10116, datetime.date(1994, 2, 27), datetime.date(1995, 12, 3), 2), (10124, datetime.date(1991, 9, 5), datetime.date(1999, 6, 6), 2), (10134, datetime.date(1993, 12, 2), datetime.date(9999, 1, 1), 2), (10144, datetime.date(1985, 10, 14), datetime.da

In [43]:
get_data(
    '''
    CREATE OR REPLACE VIEW v_dept_emp_latest_date AS
        SELECT
            emp_no, MAX(from_date) AS from_date, MAX(to_date) AS to_date
        FROM
            dept_emp
        GROUP BY emp_no;'''
)

[]

In [44]:
v_dept_emp_latest_date = get_data("SELECT * FROM employees.v_dept_emp_latest_date LIMIT 10;")
print(v_dept_emp_latest_date)


[(10001, datetime.date(1986, 6, 26), datetime.date(9999, 1, 1)), (10002, datetime.date(1996, 8, 3), datetime.date(9999, 1, 1)), (10003, datetime.date(1995, 12, 3), datetime.date(9999, 1, 1)), (10004, datetime.date(1986, 12, 1), datetime.date(9999, 1, 1)), (10005, datetime.date(1989, 9, 12), datetime.date(9999, 1, 1)), (10006, datetime.date(1990, 8, 5), datetime.date(9999, 1, 1)), (10007, datetime.date(1989, 2, 10), datetime.date(9999, 1, 1)), (10008, datetime.date(1998, 3, 11), datetime.date(2000, 7, 31)), (10009, datetime.date(1985, 2, 18), datetime.date(9999, 1, 1)), (10010, datetime.date(2000, 6, 26), datetime.date(9999, 1, 1))]


THIS VIEW WILL EXTRACT THE AVERAGE SALARY OF ALL MANAGERS REGISTERED IN THE DATABASE.

[]: # Language: sql
[]: # Path: notebooks/queries.ipynb

In [45]:
get_data(
    '''
    CREATE OR REPLACE VIEW v_managers_average_salary AS
        SELECT 
            ROUND(AVG(salary), 2) AS average_salary
        FROM 
            salaries s
        JOIN
            dept_manager m ON s.emp_no = m.emp_no;
    '''
)

[]

In [46]:
v_managers_average_salary = get_data("SELECT * FROM employees.v_managers_average_salary;")
print(v_managers_average_salary)

[(Decimal('66924.27'),)]


"STORED ROUTINES" ARE USED TO CREATE A FUNCTION THAT CAN BE USED IN OTHER QUERIES.

return the first 1000 rows from the table 'employees'

In [48]:
# get_data("DROP PROCEDURE IF EXISTS select_employees;")

In [ ]:
# get_data('''
#     CREATE PROCEDURE select_employees)
#     BEGIN
#         SELECT * FROM employees LIMIT 1000;
#     END;
# ''')

procedure that will provide the average salary of all employees

In [ ]:
# get_data('''
# DELIMITER $$
# CREATE PROCEDURE average_salary()
# BEGIN
# 	SELECT 
# 		AVG(salary)
#     FROM
# 		salaries;
# END$$
# DELIMITER ;
# ''')

In [ ]:
# get_data("CALL employees.average_salary();")